### PR TITLE
Update convolution and deconvolution output shape calculations

### PR DESCRIPTION
**Pull request by docrozza**
## Description ##

Fixes #414 

The output shape calculations for convolutions and deconvolution were incorrect.

- Conv2d / Conv3d - the dilation and stride shapes only used the first dimension value rather than the i-th dimension
- Conv1DTranspose / Conv2DTranspose - the padding and stride shapes only used the first dimension value rather than the i-th dimension. Also, the dilation shape was always assumed to be (1, 1) rather than the configured value

For more information on the correct mechanism, please refer to these [convolution calculations](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html) and [deconvolution calculations](https://pytorch.org/docs/stable/generated/torch.nn.ConvTranspose2d.html#torch.nn.ConvTranspose2d)

### CODE REVIEW

File: api/src/test/java/ai/djl/nn/convolutional/Conv1dTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline api/src/test/java/ai/djl/nn/convolutional/Conv1dTest.java:2]:
```suggestion
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: api/src/test/java/ai/djl/nn/convolutional/Conv1dTransposeTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline api/src/test/java/ai/djl/nn/convolutional/Conv1dTransposeTest.java:2]:
```suggestion
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: api/src/test/java/ai/djl/nn/convolutional/Conv2dTest.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline api/src/test/java/ai/djl/nn/convolutional/Conv2dTest.java:2]:
```suggestion
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: api/src/test/java/ai/djl/nn/convolutional/Conv2dTransposeTest.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline api/src/test/java/ai/djl/nn/convolutional/Conv2dTransposeTest.java:2]:
```suggestion
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: api/src/test/java/ai/djl/nn/convolutional/Conv3dTest.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline api/src/test/java/ai/djl/nn/convolutional/Conv3dTest.java:2]:
```suggestion
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: api/src/test/java/ai/djl/nn/convolutional/OutputShapeTest.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline api/src/test/java/ai/djl/nn/convolutional/OutputShapeTest.java:2]:
```suggestion
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: api/src/test/java/ai/djl/nn/convolutional/Range.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline api/src/test/java/ai/djl/nn/convolutional/Range.java:2]:
```suggestion
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: api/src/test/java/ai/djl/nn/convolutional/ShapeUtils.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline api/src/test/java/ai/djl/nn/convolutional/ShapeUtils.java:2]:
```suggestion
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: api/src/test/java/ai/djl/nn/convolutional/TestData.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline api/src/test/java/ai/djl/nn/convolutional/TestData.java:2]:
```suggestion
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

File: api/src/test/java/ai/djl/nn/convolutional/package-info.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (frankfliu) [inline api/src/test/java/ai/djl/nn/convolutional/package-info.java:2]:
```suggestion
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

**Review by stu1130 — APPROVED** [review]
LSTM @frankfliu please take a look at  the file structure

**Review by frankfliu — APPROVED** [review]
LGTM

**Review by lanking520 — APPROVED** [review]
